src/util/overloaded.h

```diff
@@ -0,0 +1,21 @@
+// Copyright (c) 2021 The Bitcoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#ifndef BITCOIN_UTIL_OVERLOADED_H
+#define BITCOIN_UTIL_OVERLOADED_H
+
+#include <optional>
+#include <utility>
+
+//! Overloaded helper for std::visit, useful to write code that switches on a
+//! variant type and triggers a compile error if there are any unhandled cases.
```

Comment:
4e11f88320b644b67db55fe737815451ca7d0681

I've tested this removing one case in `TxStateSerializedBlockHash` and it failed to compile.

Reply:
Just for reference, our existing use of `std::visit` will also throw compile errors even without this helper. I guess the goal of the helper is to having to write less code in the switch itself?

Reply:
> Just for reference, our existing use of `std::visit` will also throw compile errors even without this helper. I guess the goal of the helper is to having to write less code in the switch itself?

I think all existing uses of uses of `std::visit` use custom visitor classes which don't have access to local variables or anything. Or at least they did at the time this was written which was shortly after we enabled c++17. If it makes sense to define visitor classes for a particular variant type, this isn't really meant to be an alternative to that. This is just meant to be an alternative to using `if`/`else if` statements or a `switch` statement. I can update the comment to make it more clear, unless you're suggesting there might be a better way to use `std::visit` without this helper. In that case we could get rid of this.

Reply:
Not saying that this way is better, just saying that it will also emit compile errors even without the helper:

```diff
diff --git a/src/wallet/transaction.h b/src/wallet/transaction.h
index 7dc91bebd8..5a717e498c 100644
--- a/src/wallet/transaction.h
+++ b/src/wallet/transaction.h
@@ -8,12 +8,11 @@
 #include <amount.h>
 #include <primitives/transaction.h>
 #include <serialize.h>
-#include <wallet/ismine.h>
 #include <threadsafety.h>
 #include <tinyformat.h>
-#include <util/overloaded.h>
 #include <util/strencodings.h>
 #include <util/string.h>
+#include <wallet/ismine.h>
 
 #include <list>
 #include <variant>
@@ -85,25 +84,27 @@ static inline TxState TxStateInterpretSerialized(TxStateUnrecognized data)
 //! Get TxState serialized block hash. Inverse of TxStateInterpretSerialized.
 static inline uint256 TxStateSerializedBlockHash(const TxState& state)
 {
-    return std::visit(Overloaded{
-        [](const TxStateConfirmed& confirmed) { return confirmed.confirmed_block_hash; },
-        [](const TxStateConflicted& conflicted) { return conflicted.conflicting_block_hash; },
-        [](const TxStateInMempool& in_mempool) { return uint256::ZERO; },
-        [](const TxStateInactive& inactive) { return inactive.abandoned ? uint256::ONE : uint256::ZERO; },
-        [](const TxStateUnrecognized& unrecognized) { return unrecognized.block_hash; }
-    }, state);
+    struct Overloaded {
+        uint256 operator()(const TxStateConfirmed& confirmed) { return confirmed.confirmed_block_hash; }
+        uint256 operator()(const TxStateConflicted& conflicted) { return conflicted.conflicting_block_hash; }
+        uint256 operator()(const TxStateInMempool& in_mempool) { return uint256::ZERO; }
+        uint256 operator()(const TxStateInactive& inactive) { return inactive.abandoned ? uint256::ONE : uint256::ZERO; }
+        uint256 operator()(const TxStateUnrecognized& unrecognized) { return unrecognized.block_hash; }
+    };
+    return std::visit(Overloaded{}, state);
 }
 
 //! Get TxState serialized block index. Inverse of TxStateInterpretSerialized.
 static inline int TxStateSerializedIndex(const TxState& state)
 {
-    return std::visit(Overloaded{
-        [](const TxStateConfirmed& confirmed) { return confirmed.position_in_block; },
-        [](const TxStateConflicted& conflicted) { return -1; },
-        [](const TxStateInMempool& in_mempool) { return 0; },
-        [](const TxStateInactive& inactive) { return inactive.abandoned ? -1 : 0; },
-        [](const TxStateUnrecognized& unrecognized) { return unrecognized.index; }
-    }, state);
+    struct Overloaded {
+        int operator()(const TxStateConfirmed& confirmed) { return confirmed.position_in_block; }
+        int operator()(const TxStateConflicted& conflicted) { return -1; }
+        int operator()(const TxStateInMempool& in_mempool) { return 0; }
+        int operator()(const TxStateInactive& inactive) { return inactive.abandoned ? -1 : 0; }
+        int operator()(const TxStateUnrecognized& unrecognized) { return unrecognized.index; }
+    };
+    return std::visit(Overloaded{}, state);
 }